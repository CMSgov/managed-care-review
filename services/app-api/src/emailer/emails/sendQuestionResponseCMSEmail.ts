import { packageName as generatePackageName } from '../../../../app-web/src/common-code/healthPlanFormDataType'
import { formatCalendarDate } from '../../../../app-web/src/common-code/dateHelpers'
import { pruneDuplicateEmails } from '../formatters'
import type { EmailConfiguration, EmailData } from '..'
import type { ProgramType, Question } from '../../domain-models'
import {
    stripHTMLFromTemplate,
    renderTemplate,
    findContractPrograms,
} from '../templateHelpers'
import { submissionSummaryURL } from '../generateURLs'
import type { ContractRevisionWithRatesType } from '../../domain-models/contractAndRates'
import type { StateAnalystsEmails } from '..'
import { getQuestionRound } from '../../postgres/questionResponse'

export const sendQuestionResponseCMSEmail = async (
    contractRev: ContractRevisionWithRatesType,
    config: EmailConfiguration,
    statePrograms: ProgramType[],
    stateAnalystsEmails: StateAnalystsEmails,
    currentQuestion: Question,
    allContractQuestions: Question[]
): Promise<EmailData | Error> => {
    const { responses, division } = currentQuestion
    const latestResponse = responses[0]
    const questionRound = getQuestionRound(
        allContractQuestions,
        currentQuestion
    )

    if (questionRound instanceof Error) {
        return questionRound
    }

    let receiverEmails = [...stateAnalystsEmails, ...config.devReviewTeamEmails]
    if (division === 'DMCP') {
        // TODO: Remove placeholder after merging in Pearls work.
        // receiverEmails.push(...config.dmcpReviewEmails)
        receiverEmails.push('dmcpQAEmail@example.com')
    } else if (division === 'OACT') {
        receiverEmails.push(...config.oactEmails)
    }
    receiverEmails = pruneDuplicateEmails(receiverEmails)

    //This checks to make sure all programs contained in submission exists for the state.
    const packagePrograms = findContractPrograms(contractRev, statePrograms)
    if (packagePrograms instanceof Error) {
        return packagePrograms
    }

    const packageName = generatePackageName(
        contractRev.contract.stateCode,
        contractRev.contract.stateNumber,
        contractRev.formData.programIDs,
        packagePrograms
    )

    // TODO: Replace this function with Q&A function after merging in Pearls work
    const packageURL = submissionSummaryURL(
        contractRev.contract.id,
        config.baseUrl
    )

    const data = {
        packageName,
        submissionURL: packageURL,
        cmsDivision: division,
        stateResponseSubmitterEmail: latestResponse.addedBy.email,
        stateResponseSubmitterName: `${latestResponse.addedBy.givenName} ${latestResponse.addedBy.familyName}`,
        questionRound: questionRound,
        dateAsked: formatCalendarDate(currentQuestion.createdAt),
    }

    const result = await renderTemplate<typeof data>(
        'sendQuestionResponseCMSEmail',
        data
    )

    if (result instanceof Error) {
        return result
    } else {
        return {
            toAddresses: receiverEmails,
            sourceEmail: config.emailSource,
            replyToAddresses: [config.helpDeskEmail],
            subject: `${
                config.stage !== 'prod' ? `[${config.stage}] ` : ''
            }New Responses for ${packageName}`,
            bodyText: stripHTMLFromTemplate(result),
            bodyHTML: result,
        }
    }
}
