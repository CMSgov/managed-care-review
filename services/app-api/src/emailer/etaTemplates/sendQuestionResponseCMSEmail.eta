The state submitted responses to <%= it.cmsDivision %>'s questions about <%= it.packageName %><br />
<b>Submitted by:</b> <%= it.stateResponseSubmitterName %> <a href="<%= it.stateResponseSubmitterEmail %>"><%= it.stateResponseSubmitterEmail %></a><br />
<b>Round:</b> <%= it.questionRound %><br />
<b>Questions sent on:</b> <%= it.dateAsked %><br />
<br />
<a href="<%= it.submissionURL %>">View submission Q&A</a>
