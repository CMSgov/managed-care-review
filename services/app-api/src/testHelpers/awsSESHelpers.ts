import { EmailData } from '../emailer'
import { SESServiceException } from '@aws-sdk/client-ses'

async function testSendSESEmail(
    emailData: EmailData
): Promise<string | SESServiceException> {
    console.info('SendEmailCommand: ')
    return 'Email content' + JSON.stringify(emailData)
}

export { testSendSESEmail }
