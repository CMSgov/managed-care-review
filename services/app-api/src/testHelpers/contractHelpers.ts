import { InsertContractArgsType } from '../postgres/contractAndRates/insertContract'

const createDraftContractData = (
    contractArgs?: Partial<InsertContractArgsType>
): InsertContractArgsType => {
    return {
        stateCode: contractArgs?.stateCode ?? 'MN',
        submissionType: contractArgs?.submissionType ?? 'CONTRACT_AND_RATES',
        submissionDescription:
            contractArgs?.submissionDescription ?? 'Contract 1.0',
        contractType: contractArgs?.contractType ?? 'BASE',
        programIDs: contractArgs?.programIDs ?? ['PMAP'],
        populationCovered: contractArgs?.populationCovered ?? 'MEDICAID',
        riskBasedContract: contractArgs?.riskBasedContract ?? false,
    }
}

export { createDraftContractData }
