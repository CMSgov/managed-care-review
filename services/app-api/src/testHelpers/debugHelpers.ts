import util from 'util'

const consoleLogFullData = (data: unknown) =>
    console.info(util.inspect(data, false, null, true))

export { consoleLogFullData }
