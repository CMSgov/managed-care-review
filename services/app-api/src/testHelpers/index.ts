export {
    assertAnError,
    assertAnErrorExtensions,
    assertAnErrorCode,
} from './gqlAssertions'

export { must } from './errorHelpers'

export {
    createInsertContractData,
    createContractData,
    createContractRevision,
    createDraftContractData,
} from './contractAndRates/contractHelpers'

export { getProgramsFromState, getStateRecord } from './stateHelpers'
