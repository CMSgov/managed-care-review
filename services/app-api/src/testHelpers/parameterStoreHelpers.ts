import { EmailParameterStore } from '../parameterStore'

function mockEmailParameterStoreError(error?: string): EmailParameterStore {
    const message = error || 'No store found'
    return {
        getStateAnalystsSettings: async (
            stateCodes: string[]
        ): Promise<Error> => {
            return new Error(message)
        },
        getStateAnalystsEmails: async (stateCode: string): Promise<Error> => {
            return new Error(message)
        },
        getDevReviewTeamEmails: async (): Promise<Error> => {
            return new Error(message)
        },
        getCmsReviewHelpEmail: async (): Promise<Error> => {
            return new Error(message)
        },
        getCmsRateHelpEmail: async (): Promise<Error> => {
            return new Error(message)
        },
        getCmsDevTeamHelpEmail: async (): Promise<Error> => {
            return new Error(message)
        },
        getOACTEmails: async (): Promise<Error> => {
            return new Error(message)
        },
        getDMCOEmails: async (): Promise<Error> => {
            return new Error(message)
        },
        getDMCPEmails: async (): Promise<Error> => {
            return new Error(message)
        },
        getSourceEmail: async (): Promise<Error> => {
            return new Error(message)
        },
        getHelpDeskEmail: async (): Promise<Error> => {
            return new Error(message)
        },
    }
}

const getTestStateAnalystsEmails = (stateCode: string): string[] => [
    `"${stateCode} State Analyst 1" <${stateCode}StateAnalyst1@example.com>`,
    `"${stateCode} State Analyst 2" <${stateCode}StateAnalyst2@example.com>`,
]

const getTestCmsReviewSharedEmails = (): string[] => [
    `"CMS Reviewer 1" <CMS.reviewer.1@example.com>`,
    `"CMS Reviewer 2" <CMS.reviewer.2@example.com>`,
    `"CMS Reviewer 3" <CMS.reviewer.3@example.com>`,
]

const getTestOactEmails = (): string[] => [
    `"Rate Submission Reviewer 1" <rate.reviewer.1@example.com>`,
    `"Rate Submission Reviewer 2" <rate.reviewer.2@example.com>`,
    `"Rate Submission Reviewer 3" <rate.reviewer.3@example.com>`,
]

export {
    mockEmailParameterStoreError,
    getTestStateAnalystsEmails,
    getTestCmsReviewSharedEmails,
    getTestOactEmails,
}
