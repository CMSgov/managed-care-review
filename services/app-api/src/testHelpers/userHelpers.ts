import type { InsertUserArgsType } from '../postgres'
import { isStoreError, NewPostgresStore } from '../postgres'
import type {
    AdminUserType,
    CMSUserType,
    StateUserType,
    UserType,
} from '../domain-models'
import { sharedTestPrismaClient } from './storeHelpers'
import { v4 as uuidv4 } from 'uuid'

const testCMSUser = (userData?: Partial<CMSUserType>): CMSUserType => ({
    id: uuidv4(),
    role: 'CMS_USER',
    email: 'zuko@example.com',
    familyName: 'Zuko',
    givenName: 'Prince',
    divisionAssignment: 'DMCO' as const,
    stateAssignments: [],
    ...userData,
})

const testStateUser = (userData?: Partial<StateUserType>): StateUserType => ({
    id: uuidv4(),
    stateCode: 'FL',
    role: 'STATE_USER',
    email: 'james@example.com',
    familyName: 'Brown',
    givenName: 'James',
    ...userData,
})

const testAdminUser = (userData?: Partial<AdminUserType>): AdminUserType => ({
    id: uuidv4(),
    role: 'ADMIN_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const createDBUsersWithFullData = async (
    users: UserType[]
): Promise<UserType[]> => {
    const prismaClient = await sharedTestPrismaClient()
    const postgresStore = NewPostgresStore(prismaClient)

    const usersSeed: InsertUserArgsType[] = users.map((user: UserType) => {
        const userArgs: InsertUserArgsType & { id?: string } = {
            ...user,
            userID: user.id,
        }

        delete userArgs.id

        return userArgs as InsertUserArgsType
    })

    const result = await postgresStore.insertManyUsers(usersSeed)

    if (isStoreError(result)) {
        throw new Error(result.message)
    }

    return result
}

export { testAdminUser, testStateUser, testCMSUser, createDBUsersWithFullData }
