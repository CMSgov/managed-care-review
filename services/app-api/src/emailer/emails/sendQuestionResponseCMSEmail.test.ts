import {
    mockContractRev,
    mockMNState,
    mockQuestionAndResponses,
    testEmailConfig,
} from '../../testHelpers/emailerHelpers'
import { getTestStateAnalystsEmails } from '../../testHelpers/parameterStoreHelpers'
import { testCMSUser } from '../../testHelpers/userHelpers'
import { sendQuestionResponseCMSEmail } from './sendQuestionResponseCMSEmail'

const stateAnalysts = getTestStateAnalystsEmails('FL')
const oactCMSUser = testCMSUser({
    givenName: 'Bob',
    familyName: 'Smith',
    divisionAssignment: 'OACT',
})
const dmcpUser = testCMSUser({
    givenName: 'Bob',
    familyName: 'Smith',
    divisionAssignment: 'DMCP',
})
const contractRev = mockContractRev()
contractRev.formData.riskBasedContract = true
const defaultMNStatePrograms = mockMNState().programs
const questions = [
    mockQuestionAndResponses({
        id: 'test-question-id-4',
        createdAt: new Date('02/03/2024'),
        contractID: contractRev.id,
        addedBy: oactCMSUser,
        division: 'OACT',
    }),
    mockQuestionAndResponses({
        id: 'test-question-id-3',
        createdAt: new Date('02/03/2024'),
        contractID: contractRev.id,
        addedBy: dmcpUser,
        division: 'DMCP',
    }),
    mockQuestionAndResponses({
        id: 'test-question-id-2',
        createdAt: new Date('02/03/2024'),
        contractID: contractRev.id,
        addedBy: dmcpUser,
        division: 'DMCO',
    }),
    mockQuestionAndResponses({
        id: 'test-question-id-1',
        createdAt: new Date('01/03/2024'),
        contractID: contractRev.id,
        addedBy: oactCMSUser,
        division: 'OACT',
    }),
]

test.each([
    {
        questions,
        currentQuestion: mockQuestionAndResponses({
            id: 'test-question-id-4',
            createdAt: new Date('02/03/2024'),
            contractID: contractRev.id,
            addedBy: oactCMSUser,
            division: 'OACT',
        }),
        expectedResult: [
            ...stateAnalysts,
            ...testEmailConfig().devReviewTeamEmails,
            ...testEmailConfig().oactEmails,
        ],
        testDescription: 'OACT Q&A response email contains correct recipients',
    },
    {
        questions,
        currentQuestion: mockQuestionAndResponses({
            id: 'test-question-id-3',
            createdAt: new Date('02/03/2024'),
            contractID: contractRev.id,
            addedBy: dmcpUser,
            division: 'DMCP',
        }),
        expectedResult: [
            ...stateAnalysts,
            ...testEmailConfig().devReviewTeamEmails,
            ...testEmailConfig().dmcpReviewEmails,
        ],
        testDescription: 'DMCP Q&A response email contains correct recipients',
    },
    {
        questions,
        currentQuestion: mockQuestionAndResponses({
            id: 'test-question-id-2',
            createdAt: new Date('02/03/2024'),
            contractID: contractRev.id,
            addedBy: dmcpUser,
            division: 'DMCO',
        }),
        expectedResult: [
            ...stateAnalysts,
            ...testEmailConfig().devReviewTeamEmails,
        ],
        testDescription: 'DMCO Q&A response email contains correct recipients',
    },
])(
    '$testDescription',
    async ({ questions, currentQuestion, expectedResult }) => {
        const result = await sendQuestionResponseCMSEmail(
            contractRev,
            testEmailConfig(),
            defaultMNStatePrograms,
            stateAnalysts,
            currentQuestion,
            questions
        )

        if (result instanceof Error) {
            throw new Error(`Unexpect error: ${result.message}`)
        }

        expect(result).toEqual(
            expect.objectContaining({
                toAddresses: expect.arrayContaining(expectedResult),
            })
        )
    }
)

test('renders overall CMS email for a new state response as expected', async () => {
    const currentQuestion = questions[0]

    const result = await sendQuestionResponseCMSEmail(
        contractRev,
        testEmailConfig(),
        defaultMNStatePrograms,
        stateAnalysts,
        currentQuestion,
        questions
    )

    if (result instanceof Error) {
        throw new Error(`Unexpect error: ${result.message}`)
    }

    expect(result.bodyHTML).toMatchSnapshot()
})
