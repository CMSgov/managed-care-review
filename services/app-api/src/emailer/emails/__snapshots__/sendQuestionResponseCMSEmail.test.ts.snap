// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders overall CMS email for a new state response as expected 1`] = `
"The state submitted responses to OACT's questions about MCR-MN-0003-SNBC<br />
<b>Submitted by:</b> James Brown <a href="james@example.com">james@example.com</a><br />
<b>Round:</b> 2<br />
<b>Questions sent on:</b> 02/02/2024<br />
<br />
<a href="http://localhost/submissions/12345/question-and-answers">View submission Q&A</a>
"
`;
